In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
pd.set_option("display.max_columns" , None)
pd.options.display.max_rows = 10000

In [3]:
df = pd.read_csv(r'C:\Users\Vatsal\Desktop\projects\Data Sets\Self\zomato-bangalore-restaurants\zomato.csv')
print(df.shape)
df.head()

(51717, 17)


url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

In [4]:
new_df = pd.DataFrame(df)

In [5]:
print(new_df.apply(lambda new_df:len(new_df.unique())))

url                            51717
address                        11495
name                            8792
online_order                       2
book_table                         2
rate                              65
votes                           2328
phone                          14927
location                          94
rest_type                         94
dish_liked                      5272
cuisines                        2724
approx_cost(for two people)       71
reviews_list                   22513
menu_item                       9098
listed_in(type)                    7
listed_in(city)                   30
dtype: int64


In [6]:
new_df.isnull().sum()

url                                0
address                            0
name                               0
online_order                       0
book_table                         0
rate                            7775
votes                              0
phone                           1208
location                          21
rest_type                        227
dish_liked                     28078
cuisines                          45
approx_cost(for two people)      346
reviews_list                       0
menu_item                          0
listed_in(type)                    0
listed_in(city)                    0
dtype: int64

In [7]:
new_df.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'reviews_list', 'menu_item',
       'listed_in(type)', 'listed_in(city)'],
      dtype='object')

In [8]:
new_df.drop(['url','address','phone', 'location', 'dish_liked','cuisines','reviews_list','menu_item']
            , axis = 1, inplace = True)
print(new_df.shape)
display(new_df.head(5))

(51717, 9)


name online_order book_table   rate  votes  \
0                  Jalsa          Yes        Yes  4.1/5    775   
1         Spice Elephant          Yes         No  4.1/5    787   
2        San Churro Cafe          Yes         No  3.8/5    918   
3  Addhuri Udupi Bhojana           No         No  3.7/5     88   
4          Grand Village           No         No  3.8/5    166   

             rest_type approx_cost(for two people) listed_in(type)  \
0        Casual Dining                         800          Buffet   
1        Casual Dining                         800          Buffet   
2  Cafe, Casual Dining                         800          Buffet   
3          Quick Bites                         300          Buffet   
4        Casual Dining                         600          Buffet   

  listed_in(city)  
0    Banashankari  
1    Banashankari  
2    Banashankari  
3    Banashankari  
4    Banashankari

In [9]:
print(new_df.apply(lambda new_df:len(new_df.unique())))

name                           8792
online_order                      2
book_table                        2
rate                             65
votes                          2328
rest_type                        94
approx_cost(for two people)      71
listed_in(type)                   7
listed_in(city)                  30
dtype: int64


In [10]:
new_df.isnull().sum()

name                              0
online_order                      0
book_table                        0
rate                           7775
votes                             0
rest_type                       227
approx_cost(for two people)     346
listed_in(type)                   0
listed_in(city)                   0
dtype: int64

In [11]:
new_df.dropna(subset=['rest_type','approx_cost(for two people)'], axis=0, inplace=True)

We dropped missing values of variable name: 'rest_type' and 'approx_cost'because they fell in the 15% cut-off mark for dropping missing row values.
In this case: 15% of 51717 = 7757.55 or 7757 rows

In [12]:
new_df.isna().sum()

name                              0
online_order                      0
book_table                        0
rate                           7626
votes                             0
rest_type                         0
approx_cost(for two people)       0
listed_in(type)                   0
listed_in(city)                   0
dtype: int64

In [13]:
new_df.shape

(51167, 9)

In [14]:
new_df.dropna(subset=['rate'], axis=0, inplace=True)
print(new_df.shape)

(43541, 9)


In [15]:
new_df.isnull().sum()

name                           0
online_order                   0
book_table                     0
rate                           0
votes                          0
rest_type                      0
approx_cost(for two people)    0
listed_in(type)                0
listed_in(city)                0
dtype: int64

We dropped missing values of variable name: 'rate'because it fell in the 15% cut-off mark for dropping missing row values.
In this case: 15% of 51167 = 7665.05 or 7665 rows

In [16]:
new_df.dtypes

name                           object
online_order                   object
book_table                     object
rate                           object
votes                           int64
rest_type                      object
approx_cost(for two people)    object
listed_in(type)                object
listed_in(city)                object
dtype: object

In [17]:
#Filter categorical variables and get dataframe will all strings columns names
category_cols = new_df.select_dtypes(include=['object'])
#Print frequency of categories
for col in category_cols:
    #Number of times each value appears in the column
    frequency = new_df[col].value_counts()
    print('\nThis is the frequency distribution for ' + col + ':')
    print(frequency)


This is the frequency distribution for name:
Cafe Coffee Day                                                                                                                                                    89
Onesta                                                                                                                                                             85
Empire Restaurant                                                                                                                                                  71
Five Star Chicken                                                                                                                                                  68
Just Bake                                                                                                                                                          68
Kanti Sweets                                                                                                                

This is the frequency distribution for approx_cost(for two people):
300      5715
400      5541
500      4298
200      3518
600      3344
250      2283
800      2202
700      1864
1,000    1562
150      1421
350      1410
450      1265
1,200     979
1,500     947
750       749
650       748
550       715
100       702
900       677
1,300     515
1,100     510
1,400     473
2,000     356
1,600     266
1,700     247
1,800     203
3,000     162
850       162
2,500     146
2,200      78
1,900      70
2,100      67
950        62
2,800      45
4,000      29
3,500      25
2,400      23
1,350      18
180        17
3,400      13
2,300      11
2,600      10
230        10
1,250       9
40          8
130         8
50          6
1,650       6
1,450       5
330         4
1,050       4
80          4
199         4
4,100       4
2,700       3
70          3
6,000       2
240         2
120         2
3,200       2
4,500       2
360         1
3,700       1
160         1
560         1
5,000       1
Name: ap

In [18]:
print(new_df.apply(lambda new_df:len(new_df.unique())))

name                           7116
online_order                      2
book_table                        2
rate                             64
votes                          2323
rest_type                        87
approx_cost(for two people)      66
listed_in(type)                   7
listed_in(city)                  30
dtype: int64


In [19]:
new_df.rate.replace(to_replace=['1.8/5','1.8 /5'], value = '1.5', inplace=True)
new_df.rate.replace(to_replace=['2.0/5','2.1/5','2.2/5','2.3/5','2.4/5',
                               '2.0 /5','2.1 /5','2.2 /5','2.3 /5','2.4 /5'], value = '2' , inplace=True)
new_df.rate.replace(to_replace=['2.5/5','2.6/5','2.7/5','2.8/5','2.9/5',
                                '2.9 /5','2.8 /5','2.7 /5','2.6 /5', '2.5 /5'], value = '2.5' , inplace=True)
new_df.rate.replace(to_replace=['3.0/5','3.1/5','3.2/5','3.3/5','3.4/5',
                               '3.0 /5','3.1 /5','3.2 /5','3.3 /5','3.4 /5'], value = '3' , inplace=True)
new_df.rate.replace(to_replace=['3.5/5','3.6/5','3.7/5','3.8/5','3.9/5',
                                '3.9 /5','3.8 /5','3.7 /5','3.6 /5', '3.5 /5'], value = '3.5' , inplace=True)
new_df.rate.replace(to_replace=['4.0/5','4.1/5','4.2/5','4.3/5','4.4/5',
                               '4.0 /5','4.1 /5','4.2 /5','4.3 /5','4.4 /5'], value = '4' , inplace=True)
new_df.rate.replace(to_replace=['4.5/5','4.6/5','4.7/5','4.8/5','4.9/5',
                                '4.9 /5','4.8 /5','4.7 /5','4.6 /5', '4.5 /5'], value = '4.5' , inplace=True)
new_df.rate.replace(to_replace=['-','NEW'], value = '0', inplace=True)

In [20]:
new_df['rate'].value_counts()

3.5    17639
4      11057
3       9115
0       2270
2.5     2033
4.5     1244
2        178
1.5        5
Name: rate, dtype: int64

###### Break up of Rate:
1. 0 = New restaurants (Unrated)
2. 1.5 = Restaurants that got rating above 1.5 and below 2
3. 2 = Restaurants that got rating above 2 and below 2.5
4. 2.5 = Restaurants that got rating above 2.5 and below 3
5. 3 = Restaurants that got rating above 3 and below 3.5
6. 3.5 = Restaurants that got rating above 3.5 and below 4
7. 4 = Restaurants that got rating above 4 and below 4.5
8. 4.5 = Restaurants that got rating above 4.5 and below 5

In [21]:
colname =[]
for j in new_df.columns[:]:
    if new_df[j].dtype == 'object':
        colname.append(j)
colname

['name',
 'online_order',
 'book_table',
 'rate',
 'rest_type',
 'approx_cost(for two people)',
 'listed_in(type)',
 'listed_in(city)']

In [22]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()

for j in colname:
    new_df[j] = le.fit_transform(new_df[j].astype(str))

In [23]:
new_df.columns

Index(['name', 'online_order', 'book_table', 'rate', 'votes', 'rest_type',
       'approx_cost(for two people)', 'listed_in(type)', 'listed_in(city)'],
      dtype='object')

In [24]:
X1 = pd.DataFrame(new_df)
Y1 = pd.DataFrame(new_df['rate'])
print('X1 shape: ', X1.shape)
print('Y1 shape: ',Y1.shape)

X1 shape:  (43541, 9)
Y1 shape:  (43541, 1)


In [25]:
X = X1.values[:,:-1]
Y = Y1.values[:,-1]
Y = Y.astype(int)

In [26]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X)

X = scaler.transform(X)
print(X)

[[-0.24664094  0.73719724  2.40174557 ... -0.83263439  1.24124408
  -2.40511851]
 [ 1.04509188  0.73719724 -0.41636384 ... -0.83263439  1.24124408
  -2.40511851]
 [ 0.84510033  0.73719724 -0.41636384 ... -1.02361238  1.24124408
  -2.40511851]
 ...
 [-1.3538405  -1.35648908 -0.41636384 ... -1.52015515 -1.7308254
   2.73899374]
 [-1.01327519 -1.35648908  2.40174557 ... -1.52015515 -0.72144331
   2.73899374]
 [ 1.42381982 -1.35648908 -0.41636384 ... -1.48195956 -1.7308254
   2.73899374]]


In [27]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.3, random_state = 10)

## Logistic Regression:

In [28]:
#Building the model
from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression() # creating the model object
#Fitting the training data in the model
classifier.fit(X_train,Y_train)
Y_pred = classifier.predict(X_test)
#print(list(zip(Y_test,Y_pred)))
#print(classifier.coef_)
#print(classifier.intercept_)

In [29]:
#Evaluating the model

from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm= confusion_matrix (Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print("Accuracy of the model:",acc)

[[ 708    0    0    0    0    0    0]
 [  10    0   15   26    0    0    0]
 [   0    0   17  424  180    0    0]
 [   0    0    0  792 1914    0    0]
 [   0    0   10    0 5204   48    0]
 [   0    0    0    0    0 3327    0]
 [   0    0    0    0    0  301   87]]
Classification report: 
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       708
           2       0.00      0.00      0.00        51
           3       0.40      0.03      0.05       621
           4       0.64      0.29      0.40      2706
           5       0.71      0.99      0.83      5262
           6       0.91      1.00      0.95      3327
           7       1.00      0.22      0.37       388

    accuracy                           0.78     13063
   macro avg       0.66      0.50      0.51     13063
weighted avg       0.75      0.78      0.73     13063

Accuracy of the model: 0.7758554696470948


## Naive Bayes

In [30]:
# Naive Bayes
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, Y_train)

Y_pred_NB= classifier.predict(X_test)

# Summary of the predictions made by the classifier
print(classification_report(Y_test, Y_pred_NB))
print(confusion_matrix(Y_test, Y_pred_NB))
# Accuracy score
from sklearn.metrics import accuracy_score
print('accuracy is',accuracy_score(Y_pred_NB,Y_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       708
           2       1.00      1.00      1.00        51
           3       1.00      1.00      1.00       621
           4       1.00      1.00      1.00      2706
           5       1.00      1.00      1.00      5262
           6       1.00      1.00      1.00      3327
           7       1.00      1.00      1.00       388

    accuracy                           1.00     13063
   macro avg       1.00      1.00      1.00     13063
weighted avg       1.00      1.00      1.00     13063

[[ 708    0    0    0    0    0    0]
 [   0   51    0    0    0    0    0]
 [   0    0  621    0    0    0    0]
 [   0    0    0 2706    0    0    0]
 [   0    0    0    0 5262    0    0]
 [   0    0    0    0    0 3327    0]
 [   0    0    0    0    0    0  388]]
accuracy is 1.0


## Decision Tree

In [31]:
# Decision Tree's
from sklearn.tree import DecisionTreeClassifier

classifier = DecisionTreeClassifier()

classifier.fit(X_train, Y_train)

y_pred_DT = classifier.predict(X_test)

# Summary of the predictions made by the classifier
print(classification_report(Y_test, y_pred_DT))
print(confusion_matrix(Y_test, y_pred_DT))
# Accuracy score
from sklearn.metrics import accuracy_score
print('accuracy is',accuracy_score(y_pred_DT,Y_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       708
           2       1.00      1.00      1.00        51
           3       1.00      1.00      1.00       621
           4       1.00      1.00      1.00      2706
           5       1.00      1.00      1.00      5262
           6       1.00      1.00      1.00      3327
           7       1.00      1.00      1.00       388

    accuracy                           1.00     13063
   macro avg       1.00      1.00      1.00     13063
weighted avg       1.00      1.00      1.00     13063

[[ 708    0    0    0    0    0    0]
 [   0   51    0    0    0    0    0]
 [   0    0  621    0    0    0    0]
 [   0    0    0 2706    0    0    0]
 [   0    0    0    0 5262    0    0]
 [   0    0    0    0    0 3327    0]
 [   0    0    0    0    0    0  388]]
accuracy is 1.0
